### PR TITLE
Add float cast in GeneralizedRCNN normalize

### PR DESCRIPTION
**Pull request by Wadaboa**
This PR fixes issue #3228 by casting to float32 the input image of the `normalize` method in the `GeneralizedRCNNTransform` class.

Unit tests have been added, to ensure that passing `float32` or `uint8` image/mean/std variables does not lead to failure.

### CODE REVIEW

**Review by datumbox — COMMENTED** [review]
@Wadaboa Thanks for the PR. 

I left a couple of comments below, could you please check out?

File: test/test_models_detection_utils.py [hunk 1]
```diff
@@ -58,6 +58,32 @@ def test_transform_copy_targets(self):
...
+    def test_normalize_float(self):
+        transform = GeneralizedRCNNTransform(300, 500, torch.zeros(3), torch.ones(3))
+        image = [torch.rand(3, 200, 300)]
+        targets = [{'boxes': torch.rand(3, 4)}]
+        image_list, _ = transform(image, targets)  # noqa: F841
+        # check that original images still have float32 dtype
+        for img in image:
+            self.assertTrue(img.dtype == torch.float32)
+        # check that the resulting images have float32 dtype
+        self.assertTrue(image_list.tensors.dtype == torch.float32)
```
[the remark this answers is not in the document]
Comment (Wadaboa) [reply test/test_models_detection_utils.py:85]: What do you mean by "addition" assert? I was thinking about something like
```python
self.assertFalse(torch.any(torch.isnan(image_list.tensors)))
```

By the way, is the `test_normalize_float` function required, or does the `test_normalize_integer` suffice? I'm asking this because the `test_normalize_float` test would pass even without the fix (as you pointed out).

**Comment by Wadaboa** [comment]
I don't know if this is the right place to ask, but I would like to know why the `normalize` function (in the `GeneralizedRCNNTransform` class) is re-implemented from scratch if it is already available in `torchvision.transforms.functional.normalize`. Is there a reason or could it be substituted?

File: test/test_models_detection_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (datumbox) [reply test/test_models_detection_utils.py:85]: Right, let's put it on the integer. :)

**Comment by datumbox** [comment]
> I don't know if this is the right place to ask, but I would like to know why the `normalize` function (in the `GeneralizedRCNNTransform` class) is re-implemented from scratch if it is already available in `torchvision.transforms.functional.normalize`. Is there a reason or could it be substituted?

It's because the other normalize can't handle the targets.

File: test/test_models_detection_utils.py [hunk 2]
```diff
@@ -58,6 +58,34 @@ def test_transform_copy_targets(self):
...
+        image = [torch.randint(0, 255, (3, 200, 300), dtype=torch.uint8)]
+        targets = [{'boxes': torch.randint(0, 255, (3, 4))}]
+        image_list, _ = transform(image, targets)  # noqa: F841
+        # check that original images still have uint8 dtype
+        for img in image:
+            self.assertTrue(img.dtype == torch.uint8)
+        # check that the resulting images have float32 dtype
+        self.assertTrue(image_list.tensors.dtype == torch.float32)
+        # check that no NaN values are produced
+        self.assertFalse(torch.any(torch.isnan(image_list.tensors)))
```
Comment (datumbox) [inline test/test_models_detection_utils.py:76]: @Wadaboa I'm still unable to see this test failing on master without your mitigation. We need to reproduce the problem, so that we are certain that we provide the right fix. Let me know your thoughts.

Comment (Wadaboa) [reply test/test_models_detection_utils.py:76]: I just went through the code that initially produced the issue for me and I realized the following: the mentioned error can only happen when the input image has `uint8` dtype (so, in [0, 255] range) and the mean/std parameters are instead in [0, 1]. In this way, the output given by normalization will be all `inf` (not NaN as shown in the tests). 

But in this case image and mean/std would have different ranges. So, is it something that should still be checked? Or maybe I reported an error which should alert the user to check its input and this PR is not needed?

Comment (datumbox) [reply test/test_models_detection_utils.py:76]: Thanks for investigating. I agree, there are too many corner-cases to cover for uint8 inputs. This specific piece of code generally expects the input image to be floating type, so if it's not it might be worth throwing an error. 

I think what we should do is close this PR to keep things clean and open a new one where you put a type check on top of normalize and throw an exception if the input is not floating point. What you think?

Comment (Wadaboa) [reply test/test_models_detection_utils.py:76]: I agree. I'll close this PR and open a new one for the type check. Thank you.
